diff --git a/src/runtime.ts b/src/runtime.ts
index 9465d6f..fe19573 100644
--- a/src/runtime.ts
+++ b/src/runtime.ts
@@ -62,7 +62,16 @@ export class AwsV4Signer {
 }
 
 export class Configuration {
-    constructor(private configuration: ConfigurationParameters = {}) {}
+    constructor(private configuration: ConfigurationParameters = {}) {
+        if (typeof this.configuration.headers === "undefined") {
+            this.configuration.headers = {}
+        }
+
+        const headers = this.configuration.headers
+        if (!("User-Agent" in headers)) {
+            headers["User-Agent"] = "osc-sdk-js/##SDK_VERSION##"
+        }
+    }
 
     set config(configuration: Configuration) {
         this.configuration = configuration;
